### PR TITLE
Create tests for expression to cfg conversion

### PR DESCRIPTION
**Pull request by kamgal00**

### CODE REVIEW

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:78]:
```suggestion
                finalCallGraph[nameToFunMap[name]!!] = referenceSetOf(nameToFunMap[name]!!)
```

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:91]: Please split this into more lines

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:82]:
```suggestion
            functionDetailsGenerators = (nameToFunMap.values union setOf(currentFunction)).associate{ it to TestFunctionDetailsGenerator(it)}
```

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:79]: This can be simplified to two `associate`s (example usage of `associate` below)

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:83]:
```suggestion
            for ((name, variables) in funToAffectedVar) {
```

Comment (ignacy123) [reply app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:83]: Also, I think `. forEach { }` construct might be better suited here

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:147]:
```suggestion
        fun <T> ReferenceHashMap<T, T>.ensurePairSymmetrical(a: T, b: T): Boolean {
```
The previous name suggests this function has no side-effects

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:173]:
```suggestion
         (this.args zip iftNode.args).forEach { if // logic }
```

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:190]: We can `zip` here as well

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:613]:
```suggestion
        val cfg = context.createCfg( //   x + ( g( a + f() , b ) + ( ( b + x ) + ( h() ? x : y ) ) ) f -> x, h -> b
```

**Review by ignacy123 — CHANGES_REQUESTED** [review]
The tests themselves seem quite exhaustive already, but I'd suggest adding:
 - some tests for all basic literals (literals and unary operations). They might be helpful if there will be some problems with testing framework/CFG calculations very early on.
 - some complex tests with functions not affecting variables at the same level. This is to ensure that the CFG builder can take advantage of that and allows for parallel processing.

Most of my remarks are related to style and decreasing line count. Some of them are also trying to prevent code duplication with other branches before it occurs - please try to incorporate these suggestions.
I am a fan of your usage of `infix` notation, really helps with readability :+1:

**Comment by bebidek** [comment]
Should be able to auto-merge to master now.

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [reply app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:80]: won't work cause reference maps

File: app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate_form/ExpressionControlFlowTest.kt:308]: This shouldn't be here I think :stuck_out_tongue:

**Review by ignacy123 — CHANGES_REQUESTED** [review]
I left some comments in the [old PR](https://github.com/TCS-Compilers/compiler/pull/84), which got merged before my approval.
Please address them here, as this is the only place where this can be done now.

**Review by ignacy123 — APPROVED** [review]

**Pull request merged**
